object.c

```diff
@@ -4065,72 +4065,77 @@ rb_f_loop_size(VALUE self, VALUE args, VALUE eobj)
- *  from the root like <code>::File</code> or <code>::Enumerator</code>.
- *  Like #method_missing, #const_missing can be used to delegate constant
- *  lookup to +Object+:
- *
- *    class MyObjectSystem < BasicObject
- *      def self.const_missing(name)
- *        ::Object.const_get(name)
+/*
+ *  Document-class: BasicObject
+ *
+ *  :markup: markdown
+ *
+ *  `BasicObject` is the parent class of all classes in Ruby.
+ *  In particular, `BasicObject` is the parent class of class Object,
+ *  which is itself the default parent class of every Ruby class:
+ *
+ *  ```
+ *  class Foo; end
+ *  Foo.superclass    # => Object
+ *  Object.superclass # => BasicObject
+ *  ```
+ *
+ *  `BasicObject` is the only class that has no parent:
+ *
+ *  ```
+ *  BasicObject.superclass # => nil
+ *  ```
+ *
+ *  \Class `BasicObject` can be used to create an object hierarchy
+ *  (e.g., class Delegator) that is independent of Ruby's object hierarchy.
+ *  Such objects:
+ *
+ *  - Do not have namespace "pollution" from the many methods
+ *    provided in class Object and its included module Kernel.
+ *  - Do not have definitions of common classes,
+ *    and so references to such common classes must be fully qualified
+ *    (`::String`, not `String`).
+ *
+ *  A variety of strategies can be used to provide useful portions
+ *  of the Standard Library in subclasses of `BasicObject`:
+ *
+ *  - The immediate subclass could `include Kernel`,
+ *    which would define methods such as `puts`, `exit`, etc.
+ *  - A custom Kernel-like module could be created and included.
+ *  - Delegation can be used via #method_missing:
+ *
+ *      ```
+ *      class MyObjectSystem < BasicObject
+ *        DELEGATE = [:puts, :p]
+ *        def method_missing(name, *args, &block)
+ *          return super unless DELEGATE.include? name
+ *          ::Kernel.send(name, *args, &block)
+ *        end
+ *        def respond_to_missing?(name, include_private = false)
+ *          DELEGATE.include?(name)
+ *        end
  *      end
- *    end
+ *      ```
  *
- *  === What's Here
+ *  ## What's Here
  *
  *  These are the methods defined for \BasicObject:
  *
- *  - ::new: Returns a new \BasicObject instance.
- *  - #!: Returns the boolean negation of +self+: +true+ or +false+.
- *  - #!=: Returns whether +self+ and the given object are _not_ equal.
- *  - #==: Returns whether +self+ and the given object are equivalent.
- *  - #__id__: Returns the integer object identifier for +self+.
- *  - #__send__: Calls the method identified by the given symbol.
- *  - #equal?: Returns whether +self+ and the given object are the same object.
- *  - #instance_eval: Evaluates the given string or block in the context of +self+.
- *  - #instance_exec: Executes the given block in the context of +self+,
- *    passing the given arguments.
+ *  | \\Method                    | Effect                                                                          |
```

Comment:
Why are we switching from bullet points to tables? Are we planning to change all "What's Here" sections from bullet points to tables?

Reply:
The GFM table is, IMO, *much* more readable.  Whether all such should be changed is a matter for discussion.  (Plus several other issues around "What's Here" sections.)  Where's a good place to start such a discussion?

Issues:
- Should the section have a GFM instead of bullets?
- Should attributes be included (in addition to methods defined with `def`)?  Why or why not?
- Should the entire section be off-page (in a separate linked file)?  That would make it simple to do a GFM table in a separate `.md` file (instead of changing a whole RDoc section from markup to markdown).  If so, where would those files live?
- Should there be a discussion and example of an included or inherited method that may behave in an unusual way?  (I'm thinking of one that calls into a local method in the local class, which may make behavior differ from class to class.)
- Should constants be included?  I think, *emphatically*, No;  but if not, should the title be changed from "What's Here" to something about methods?

Reply:
> Where's a good place to start such a discussion?

I don't think there's a formal place to start a discussion, but you can try bugs.ruby-lang.org if you want. We can have some discussions here for now.

> Should the section have a GFM instead of bullets?

As-is, I don't think the table is any more readable than the bullet points. Do you plan on adding additional columns to the table? If so, what do you plan on adding? We should be cautious in adding too much information in the "What's Here" section in order to not bloat it and turn it from a quick reference into method documentation.

> Should attributes be included (in addition to methods defined with def)? Why or why not?

I'm not sure what you mean by attributes? Do you mean constants or method parameters? 

> Should the entire section be off-page (in a separate linked file)? That would make it simple to do a GFM table in a separate .md file (instead of changing a whole RDoc section from markup to markdown). If so, where would those files live?

I don't think it should live in a separate file. Documentation that live in a separate file should either be shared information that is used in multiple places or information that is not critical. I don't think the "What's Here" section will ever be shared across multiple files. If we place the "What's Here" section in another file and link it, then it makes it harder to use because the user can no longer search on the page and it's another level of indirection the user needs to do to access the information.

> Should there be a discussion and example of an included or inherited method that may behave in an unusual way? (I'm thinking of one that calls into a local method in the local class, which may make behavior differ from class to class.)

IMO, the "What's Here" section should be simple. I use it to find a method that I think exists but am unsure. I quickly skim the description to find candidates that are promising. The more complicated it is, the less useful it is for me. If I need examples, I expect to find it in the method documentation.

> Should constants be included? I think, emphatically, No; but if not, should the title be changed from "What's Here" to something about methods?

We could consider adding a section for just constants, or just putting the constants at the top of the list. But IMO, most constants are supplementary to the methods, meaning that users use the method and use constants to help them use the methods. In this case, I don't think the constant should be referenced in the "What's Here" section and instead only be referenced in the method documentation.

Reply:
Mostly agree with you, @peterzhu2118.

- Exception:  as I said, I think the table is much more readable; maybe others will have opinions?
- By "attributes," I mean methods defined by `attr_*`.  These appear to the user as full-fledged methods, though some say that internally they're different from those defined by `def`.
- Separate file is not a good idea.
- Discussion outside of the table complicates things, and should be avoided.
- Constants should not be included.

Reply:
> By "attributes," I mean methods defined by attr_*. These appear to the user as full-fledged methods, though some say that internally they're different from those defined by def.

Ah attribute accessors. Yes these should definitely be included since they (for the most part) behave identically to methods (and the user should treat it like a method anyways).

Reply:
> As-is, I don't think the table is any more readable than the bullet points.

I'm for @peterzhu2118.

> No plan to add columns.

If you just want to align columns, why not definition list?